File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
class Solution {
public:
    int getSize(ListNode* head){
        int size = 0;
        for(ListNode* start = head; start != NULL; start = start->next){
            size++;
        }
        return size;
    }
    
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int sizeOfList = getSize(head);
        int k = sizeOfList - n;
        
        if(sizeOfList == 1){
            return NULL;
        }
        
        ListNode* newNode = new ListNode(-1);
        ListNode* prevNode = newNode;
        ListNode* currNode = head;
        
        while(currNode){
            if(k == 0){
                break;
            }
            
            prevNode->next = currNode;
            prevNode = currNode;
            currNode = currNode->next;
            
            k--;
        }
        
        prevNode->next = currNode->next;
        
        return newNode->next;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode right = head;
        ListNode left = head;
        for(int i = 0; i < n; i++){
            right = right.next;
        }
        
        if(right == null){
            return head.next;
        }
        
        while(right.next != null){
            left = left.next;
            right = right.next;
        }
        
        left.next = left.next.next;
        
        return head;
    }
}